File: docs/js/blackjack/rules.js
export const Action = {
  HIT: "HIT", STAND: "STAND", DOUBLE: "DOUBLE", SPLIT: "SPLIT", SURRENDER: "SURRENDER",
};

export const MAX_SPLITS = 3;

export function dealerShouldHit(hand, dealerHitsSoft17) {
  const value = hand.value;
  if (value < 17) return true;
  if (value === 17 && hand.isSoft && dealerHitsSoft17) return true;
  return false;
}

export function dealerShouldPeek(upCard) {
  return upCard.isAce || upCard.isTenValue;
}

export function legalActions(hand, state, config, isFirstAction, playerBankroll) {
  if (hand.isFinished || hand.isSurrendered || hand.isBust) return new Set();

  const actions = new Set([Action.HIT, Action.STAND]);

  if (isFirstAction && state.allowSurrender && !hand.isFromSplitAces) {
    actions.add(Action.SURRENDER);
  }

  const canDouble = hand.cards.length === 2 && !hand.isDoubled;
  if (canDouble && playerBankroll >= hand.bet) actions.add(Action.DOUBLE);

  if (hand.canSplit(MAX_SPLITS) && playerBankroll >= hand.bet) actions.add(Action.SPLIT);

  if (hand.isFromSplitAces && hand.cards.length >= 2) {
    actions.delete(Action.HIT);
    actions.delete(Action.DOUBLE);
    actions.delete(Action.SPLIT);
  }

  return actions;
}

export function createTableState(dealerUpCard) {
  return {
    dealerUpCard,
    dealerHasBlackjack: false,
    insuranceOffered: false,
    allowSurrender: true,
    allowDoubleAfterSplit: true,
  };
}

File: blackjack/rules.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto

from blackjack.cards import Card
from blackjack.config import GameConfig
from blackjack.hand import Hand


class Action(Enum):
    HIT = auto()
    STAND = auto()
    DOUBLE = auto()
    SPLIT = auto()
    SURRENDER = auto()
    INSURANCE = auto()


class RoundPhase(Enum):
    BETTING = auto()
    DEAL = auto()
    INSURANCE = auto()
    PLAYER_TURNS = auto()
    DEALER_TURN = auto()
    SETTLEMENT = auto()
    COMPLETE = auto()


MAX_SPLITS = 3


@dataclass
class TableState:
    dealer_up_card: Card | None
    dealer_has_blackjack: bool = False
    insurance_offered: bool = False
    allow_surrender: bool = True
    allow_double_after_split: bool = True


def dealer_should_hit(hand: Hand, config: GameConfig) -> bool:
    value = hand.value
    if value < 17:
        return True
    if value == 17 and hand.is_soft and config.dealer_hits_soft_17:
        return True
    return False


def dealer_should_peek(up_card: Card) -> bool:
    return up_card.is_ace or up_card.is_ten_value


def legal_actions(
    hand: Hand,
    state: TableState,
    config: GameConfig,
    *,
    is_first_action: bool,
    player_bankroll: int,
    num_active_hands: int,
) -> set[Action]:
    if hand.is_finished or hand.is_surrendered or hand.is_bust:
        return set()

    actions: set[Action] = {Action.HIT, Action.STAND}

    if is_first_action and state.allow_surrender and not hand.is_from_split_aces:
        actions.add(Action.SURRENDER)

    can_double = len(hand.cards) == 2 and not hand.is_doubled
    if can_double and player_bankroll >= hand.bet:
        actions.add(Action.DOUBLE)

    if hand.can_split(MAX_SPLITS) and player_bankroll >= hand.bet:
        actions.add(Action.SPLIT)

    if hand.is_from_split_aces and len(hand.cards) >= 2:
        actions.discard(Action.HIT)
        actions.discard(Action.DOUBLE)
        actions.discard(Action.SPLIT)

    return actions
